internal/fields/validate.go

```diff
@@ -0,0 +1,196 @@
+
+		switch val.(type) {
+		case []map[string]interface{}:
+			for _, m := range val.([]map[string]interface{}) {
+				err = v.validateMapElement(key, m)
+				if err != nil {
+					return err
+				}
+			}
+		case map[string]interface{}:
+			err = v.validateMapElement(key, val.(map[string]interface{}))
+			if err != nil {
+				return err
+			}
+		default:
+			err = v.validateElementFormat(key, val)
+			if err != nil {
+				return err
+			}
+			break
+		}
+	}
+	return nil
+}
+
+func (v *Validator) validateElementFormat(key string, val interface{}) error {
+	if key == "" {
+		return nil // root key is always valid
+	}
+
+	if skipValidationForField(key) {
+		return nil // generic field, let's skip validation for now
+	}
+
+	definition := findElementDefinitionInSlice("", key, v.schema)
+	if definition == nil {
+		return fmt.Errorf(`field "%s" is not defined`, key)
+	}
+
+	err := parseElementValue(key, *definition, val)
+	if err != nil {
+		return errors.Wrap(err, "parsing field value failed")
+	}
+	return nil
+}
+
+func skipValidationForField(key string) bool {
+	return isFieldFamilyMatching("agent", key) ||
+		isFieldFamilyMatching("elastic_agent", key)
+}
+
+func isFieldFamilyMatching(family, key string) bool {
+	return key == family || strings.HasPrefix(key, family+".")
+}
+
+func findElementDefinitionInSlice(root, searchedKey string, fieldDefinitions []fieldDefinition) *fieldDefinition {
+	for _, def := range fieldDefinitions {
+		key := strings.TrimLeft(root+"."+def.Name, ".")
+		if compareKeys(key, def, searchedKey) {
+			return &def
+		}
+
+		if len(def.Fields) == 0 {
+			continue
+		}
+
+		fd := findElementDefinitionInSlice(key, searchedKey, def.Fields)
+		if fd != nil {
+			return fd
+		}
+	}
+	return nil
+}
+
+func compareKeys(key string, def fieldDefinition, searchedKey string) bool {
+	k := strings.ReplaceAll(key, ".", "\\.")
+	k = strings.ReplaceAll(k, "*", "[^.]+")
+
+	// Workaround for potential geo_point, as "lot" and "lat" fields are not present in field definitions.
+	if def.Type == "geo_point" {
+		k += "\\.(lon|lat)"
+	}
+
+	k = fmt.Sprintf("^%s$", k)
+	matched, err := regexp.MatchString(k, searchedKey)
+	if err != nil {
+		panic(err)
+	}
+	return matched
+}
+
+func parseElementValue(key string, definition fieldDefinition, val interface{}) error {
+	val, ok := ensureSingleElementValue(val)
+	if !ok {
+		return nil // it's an array, but it's not possible to extract the single value.
+	}
+
+	var valid bool
+	switch definition.Type {
+	case "date", "ip", "constant_keyword", "keyword", "text":
+		_, valid = val.(string)
+	case "float", "long", "double":
+		_, valid = val.(float64)
+	default:
+		valid = true // all other types are considered valid not blocking validation
+	}
+
+	if !valid {
+		return fmt.Errorf("field \"%s\" has invalid type, expected: %s, actual Go type: %s", key, definition.Type, reflect.TypeOf(val))
+	}
+	return nil
+}
+
+func ensureSingleElementValue(val interface{}) (interface{}, bool) {
```

Comment:
Took me a few seconds to realize why we need this helper: because any data type in ES can take array values. Maybe add an explanatory comment here for future us?

Reply:
I added a comment.